File: Lab2/app/src/main/java/com/example/lab2/SimpleChatActivity.java
package com.example.lab2;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.view.View;
import android.widget.Button;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.MqttCallback;
import org.eclipse.paho.client.mqttv3.MqttClient;
import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;
import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;

import java.lang.ref.WeakReference;
import java.util.ArrayList;

public class SimpleChatActivity extends AppCompatActivity {
    private String myTopic = "AGH/FrozenTear7";

    private ListView chatMessagesListView;

    private EditText messageEditText;

    String server;
    String nick;

    ArrayList<String> listItems;
    ArrayAdapter<String> adapter;

    Handler myHandler = new MyHandler(this);

    MqttClient sampleClient = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_simple_chat);

        TextView nicknameChatTextView = findViewById(R.id.nicknameChatTextView);
        chatMessagesListView = findViewById(R.id.chatMessagesListView);
        messageEditText = findViewById(R.id.messageEditText);
        Button sendButton = findViewById(R.id.sendButton);

        server = getIntent().getStringExtra(MainActivity.IP);
        nick = getIntent().getStringExtra(MainActivity.NICK);

        listItems = new ArrayList<>();
        adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, listItems);

        nicknameChatTextView.setText(getIntent().getStringExtra(MainActivity.NICK));

        chatMessagesListView.setAdapter(adapter);

        sendButton.setOnClickListener(this::postOnClick);

        new Thread(this::startMQTT).start();
    }

    public void postOnClick(View view) {
        Message msg = myHandler.obtainMessage();
        Bundle b = new Bundle();
        b.putString("NICK", nick);
        b.putString("MSG", messageEditText.getText().toString());
        msg.setData(b);

        try {
            sampleClient.publish(myTopic, new MqttMessage(messageEditText.getText().toString().getBytes()));
        } catch (MqttException e) {
            Toast.makeText(getApplicationContext(), "Could not send message", Toast.LENGTH_SHORT).show();
        }
    }

    private static class MyHandler extends Handler {
        private final WeakReference<SimpleChatActivity> sActivity;

        MyHandler(SimpleChatActivity activity) {
            sActivity = new WeakReference<>(activity);
        }

        public void handleMessage(Message msg) {
            SimpleChatActivity activity = sActivity.get();
            activity.listItems.add(msg.getData().getString("MSG"));
            activity.adapter.notifyDataSetChanged();
            activity.chatMessagesListView.setSelection(activity.listItems.size() - 1);
        }
    }

    private void startMQTT() {
        String clientId;
        MemoryPersistence persistence = new MemoryPersistence();
        try {
            String broker = "tcp://" + server + ":1883";
            clientId = nick;
            sampleClient = new MqttClient(broker, clientId, persistence);
            sampleClient.setCallback(new MqttCallback() {
                @Override
                public void connectionLost(Throwable throwable) {
                    System.out.println("Connection lost");
                }

                @Override
                public void messageArrived(String s, MqttMessage mqttMessage) {
                    System.out.println(s);
                    System.out.println(mqttMessage);
                    Message msg = myHandler.obtainMessage();
                    Bundle b = new Bundle();
                    b.putString("MSG", mqttMessage.toString());
                    msg.setData(b);
                    myHandler.sendMessage(msg);
                }

                @Override
                public void deliveryComplete(IMqttDeliveryToken iMqttDeliveryToken) {
                    System.out.println("Delivery complete");
                }
            });
            MqttConnectOptions connOpts = new MqttConnectOptions();
            connOpts.setCleanSession(true);
            System.out.println("Connecting to broker: " + broker);
            sampleClient.connect(connOpts);
            System.out.println("Connected");
            sampleClient.subscribe(myTopic);
        } catch (MqttException e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (sampleClient != null) {
            try {
                sampleClient.disconnect();
            } catch (MqttException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Lab2/app/src/main/java/com/example/lab2/SimpleChatActivity.kt
package com.example.lab2

import android.annotation.SuppressLint
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import java.lang.ref.WeakReference


class SimpleChatActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_simple_chat)

        val nicknameChatTextView = findViewById<TextView>(R.id.nicknameChatTextView)

        val chatMessagesListView = findViewById<ListView>(R.id.chatMessagesListView)

        val messageEditText = findViewById<EditText>(R.id.messageEditText)

        val sendButton = findViewById<Button>(R.id.sendButton)

        val server = intent.getStringExtra(MainActivity.IP)
        val nick = intent.getStringExtra(MainActivity.NICK)

        nicknameChatTextView.text = nick

        val listItems = ArrayList<String>()
        val adapter: ArrayAdapter<String>
        adapter = ArrayAdapter(this, android.R.layout.simple_list_item_1, listItems)
        chatMessagesListView.adapter = adapter

        val myHandler: Handler = object : Handler() {
            @SuppressLint("HandlerLeak")
            override fun handleMessage(msg: Message) {
                listItems.add(
                    "[" + msg.data.getString("NICK").toString() + "]" +
                            msg.data.getString("MSG")
                )
                adapter.notifyDataSetChanged()
                chatMessagesListView.setSelection(listItems.size - 1)
            }
        }

//        class MyHandler(activity: SimpleChatActivity?) : Handler() {
//            private val sActivity: WeakReference<SimpleChatActivity> =
//                WeakReference<SimpleChatActivity>(activity)
//
//            override fun handleMessage(msg: Message) {
//                val activity: SimpleChatActivity? = sActivity.get()
//                activity!!.listItems.add(
//                    "[" + msg.data.getString("NICK") + "]" +
//                            msg.data.getString("MSG")
//                )
//                activity.adapter.notifyDataSetChanged()
//                activity.chatMessagesListView.setSelection(activity.listItems.size() - 1)
//            }
//        }

//        val myHandler: Handler = MyHandler(this)

        fun postOnClick() {
            val msg: Message = myHandler.obtainMessage()
            val b = Bundle()
            b.putString("NICK", nick)
            b.putString("MSG", messageEditText.text.toString())
            msg.data = b
            myHandler.sendMessage(msg)
        }

        sendButton.setOnClickListener {
            postOnClick()
        }

        MqttClient sampleClient=null;
        private void startMQTT(){
            String clientId;
            MemoryPersistence persistence = new MemoryPersistence();
            try {
                String broker = "tcp://"+ip+":1883";
                clientId = nick;
                sampleClient = new MqttClient(broker, clientId, persistence);
                sampleClient.setCallback(new MqttCallback() {
                    //TODO
                });
                MqttConnectOptions connOpts = new MqttConnectOptions();
                connOpts.setCleanSession(true);
                System.out.println("Connecting to broker: "+broker);
                sampleClient.connect(connOpts);
                System.out.println("Connected");
                sampleClient.subscribe("#");
            } catch (MqttException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

    }
}